src/shogun/machine/gp/EPInferenceMethod.cpp

```diff
@@ -494,22 +495,28 @@ SGVector<float64_t> CEPInferenceMethod::get_derivative_wrt_inference_method(
 }
 
 SGVector<float64_t> CEPInferenceMethod::get_derivative_wrt_likelihood_model(
-		const TParameter* param)
+		const AnyParameter *param)
 {
 	SG_NOTIMPLEMENTED
 	return SGVector<float64_t>();
 }
 
 SGVector<float64_t> CEPInferenceMethod::get_derivative_wrt_kernel(
-		const TParameter* param)
+		const AnyParameter *param)
 {
 	// create eigen representation of the matrix Q
 	Map<MatrixXd> eigen_F(m_F.matrix, m_F.num_rows, m_F.num_cols);
 
 	REQUIRE(param, "Param not set\n");
+
 	SGVector<float64_t> result;
-	int64_t len=const_cast<TParameter *>(param)->m_datatype.get_num_elements();
-	result=SGVector<float64_t>(len);
+	auto param_value = param->get_value();
+	int64_t len;
+    auto f_scalar = [&len](auto value) { len=1; };
+    auto f_vector = [&len, &param_value](auto value) { len=param_value.as<SGVector<float64_t>>().vlen; };
+    auto f_matrix = [&len, &param_value](auto value) { len=param_value.as<SGMatrix<float64_t>>().num_rows*param_value.as<SGMatrix<float64_t>>().num_cols; };
+    sg_any_dispatch(param->get_value(), sg_all_typemap, f_scalar, f_vector, f_matrix);
```

Comment:
This is needed many times so maybe we could move it in Any?

Reply:
You mean the lambdas? I think it's better to leave them where they are. It's a zero cost abstraction anyway and improves readability